filebeat/docs/inputs/input-filestream-file-options.asciidoc

```diff
@@ -452,3 +452,60 @@ Set the location of the marker file the following way:
 ----
 file_identity.inode_marker.path: /logs/.filebeat-marker
 ----
+
+==== Log rotation
+
+As log files are constantly written, they must be rotated and purged to prevent
+the logger application from filling up the disk. Rotation is done by an external
+application, thus, {beatname_uc} needs information how to cooperate with it.
+
+When reading from rotating files make sure the paths configuration includes
+both the active file and all rotated files.
+
+By default, {beatname_uc} is able to track files correctly in the following strategies:
+* create: new active file with a unique name is created on rotation
+* rename: rotated files are renamed
+
+However, in case of copytruncate strategy, you should provide additional configuration
+to {beatname_uc}.
+
+[float]
+===== rotation.external.strategy.copytruncate
+
+If the log rotating application copies the contents of the active file and then
+truncates the original file, use these options to help {beatname_uc} to read files
+correctly.
+
+Set the option `suffix_regex` so {beatname_uc} can tell active and rotated files apart. There are
+two supported suffix types in the input: numberic and date.
+
+====== Numeric suffix
+
+If your rotated files have an incrementing index appended to the end of the filename, e.g.
+active file `apache.log` and the rotated files are named `apache.log.1`, `apache.log.2`, etc,
+use the following configuration.
+
+[source,yaml]
+---
+rotation.external.strategy.copytruncate:
+  suffix_regex: \.\d$
+  count: 7
+---
+
+`count` is the number of rotated files. {beatname_uc} needs this information to clean the
+registry when the rotated file has been purged.
```

Comment:
> What happens if we keep this `0`? Will files be removed based on ttl? Shall we automatically set `clean_removed` for rotated files?

0 is an invalid setting as I use this number to when creating a slice for `rotatedFiles`.

Rotated files that are removed are immediately stopped and cleaned up, so we keep number of open files as low as possible and the size of the registry as small as possible. However, I am open to keeping the files open until they stop themselves and let the registry clean the state if you think that is better.

> Regex and count seem to be helpful for other log rotation strategies (move files) as well, as it allows us to close files early and remove state from the registry early.
> Closing files on EOF reduces the number of FDs we need to keep + while removing state from registry more early helps with inode reuse issues.

Indeed, that's why I enable `close_eof` and remove the state of rotated files when they are deleted.

Reply:
> 0 is an invalid setting as I use this number to when creating a slice for rotatedFiles.

Hm, I wonder if we really the `count` setting at all. For slices we should use `append`, but we should always be ready to allow slices to grow. If someone updates the application settings to produce more logs when trying to hunt down an issue, the user should not have to remember to update/change the beats setting as well.

> Rotated files that are removed are immediately stopped and cleaned up, so we keep number of open files as low as possible and the size of the registry as small as possible. However, I am open to keeping the files open until they stop themselves and let the registry clean the state if you think that is better.

This behavior should be configurable. Normally this is what `close_removed` and `clean_removed` had been used for. The number of file descriptors can indeed grow due to bandwidth limitation, but we don't want to drops logs, just because a file was rotated and we have to wait 1min for ACK from Elasticsearch if the hole environment is low EPS only. More sophisticated alternative is to configure a limit of removed files that should be kept open. By default  we would keep open up to `2` removed files for a log-stream for example. Once the limit is reached, we delete the close and clean the files we have kept open the longest (oldest files).

Reply:
I have kept removing and cleaning configurable.

Reply:
Do you still need to update the doc? I still see `count` being mentioned.